Update documentation (#250)
xenit-eu/dynamic-extensions-for-alfresco: README.md

# Dynamic Extensions for Alfresco

[![Build Status](https://travis-ci.org/xenit-eu/dynamic-extensions-for-alfresco.svg)](https://travis-ci.org/xenit-eu/dynamic-extensions-for-alfresco)

Add OSGi based hot-deploy functionality and Spring annotation based configuration to your Alfresco.

## Introduction

Rapid development of Alfresco repository extensions in Java. Deploy your code in seconds, not minutes. Life is too short for endless server restarts.

Dynamic Extensions adds an OSGi container to the Alfresco repository, enabling live deployment of Java code, with no need to restart the server.
Alfresco itself is not "OSGi-fied" in any way; the OSGi container runs on top of the core Alfresco platform.

> Default Alfresco Platform Extensions use Spring XML based configuration. With Dynamic Extensions, developers

Reviewer: ```suggestion
Standard Alfresco Platform Extensions use Spring XML based configuration. With Dynamic Extensions, developers
```

have the ability to create Alfresco Platform Extensions using Spring's annotations based configuration.

## Example Dynamic Extensions based Alfresco Platform extension

This example Web Script examines a node and passes information to a Freemarker template:
```java
@Component
@WebScript
public ExampleWebScript {

  @Autowired
  private NodeService nodeService;

  @Uri("/show-node")
  // Example: http://localhost/alfresco/service/show-node?nodeRef=workspace://SpacesStore/12345
  public Map<String, Object> displayNodeName(@RequestParam NodeRef nodeRef) {
    Map<String, Object> model = new HashMap<String, Object>();
    model.put("properties", nodeService.getProperties(nodeRef));    
    return model; // Model is passed to Freemarker template.
  }
}
```

Note that this is an _annotation Web Script_. These types of Web Script are configured through Java annotations instead of `*.desc.xml` descriptors. Annotation Web Scripts are similar to Spring MVC's annotation-based controllers.

Here's the accompanying Freemarker template fragment:

> ```html

Reviewer: I think it needs to be mentionned where this file has to be placed.

Author: This is mentioned in the other documentation, KIS here?

Reviewer: Okay

<table>
  <#list properties?keys as name>    
    <tr>
      <th>${name}</th>
      <td>${properties[name]!''}</td>
    </tr>
  </#list>
</table>
```

This is all the code that is required; there's no need for Spring XML config or Web Script XML descriptors. Hot-reloading and reducing configuration overhead are not particularly novel concepts in the Java development world at large. Essentially, Dynamic Extensions modernizes the development of Alfresco repository extensions.

The example above may be trivial, but the point is that, behind the scenes, services are still wired together through Spring and handled by the Web Script framework. Conceptually there is no real difference between a Dynamic Extension and a regular Alfresco extension. There's just less overhead and more convenience.

## Installing Dynamic Extensions in Alfresco

Dynamic Extensions (DE) is distributed as an 
[Alfresco Module Package (AMP)](https://docs.alfresco.com/5.2/concepts/dev-extensions-packaging-techniques-amps.html) extension that can be installed in Alfresco.

> ### Installing the DE AMP

Reviewer: ```suggestion
### Installing the Dynamic Extensions AMP
```

> To support multiple Alfresco versions, different AMP's for each minor Alfresco version update are 

Reviewer: ```suggestion
To support multiple Alfresco versions, different AMPs for each minor Alfresco version update are 
```

> build and distributed. E.g. if you are working with Alfresco 6.0.7-ga, you should

Reviewer: ```suggestion
published.
E.g. if you are working with Alfresco 6.0.7-ga, you should
```

use the `alfresco-dynamic-extensions-repo-60` artifact.

#### Maven Central Coordinates
All required artifacts, including the AMP to be installed in Alfresco, are available in Maven Central.

```xml
<dependency>
    <groupId>eu.xenit</groupId>
    <artifactId>alfresco-dynamic-extensions-repo-${alfresco-version}</artifactId>
    <version>${latest-dynamic-extensions-version}</version>
    <type>amp</type>
</dependency>
```

```groovy

> alfrescoAmp "eu.xenit:alfresco-dynamic-extensions-repo-${alfrescoVersion}:${latest-dynamic-extensions-version}@amp"

Reviewer: ```suggestion
alfrescoAmp "eu.xenit:alfresco-dynamic-extensions-repo-${alfrescoVersion}:${dynamicExtensionsVersion}@amp"
```

```

> These artifacts can be used to automatically install DE in Alfresco using e.g. the Alfresco Maven SDK or 

Reviewer: ```suggestion
These artifacts can be used to automatically install Dynamic Extensions in Alfresco using e.g. the Alfresco Maven SDK or 
```

> the [Alfresco Gradle Docker Plugins](https://github.com/xenit-eu/alfresco-docker-gradle-plugin)

Reviewer: ```suggestion
the [Alfresco Docker Gradle Plugins](https://github.com/xenit-eu/alfresco-docker-gradle-plugin)
```

#### Manual download and install

* Download the latest <a href="https://github.com/xenit-eu/dynamic-extensions-for-alfresco/releases">Dynamic Extensions AMP</a>.
* Use the <a href="https://docs.alfresco.com/6.1/concepts/dev-extensions-modules-management-tool.html">Module Management Tool</a> to install the AMP in the Alfresco repository of your choosing.
* After restarting Alfresco, open the Control Panel: <a href="http://localhost:8080/alfresco/service/dynamic-extensions/">http://localhost:8080/alfresco/service/dynamic-extensions/</a>.
* Accessing the Control Panel requires an admin account.

### Supported Alfresco versions

> Dynamic Extensions is systematically integration tested against:

Reviewer 2: integration-tested

* Alfresco Enterprise 5.1
* Alfresco Enterprise 5.0

> #### Known Alfresco issues that impact Dynamic Extensions</summary>

Reviewer: ```suggestion
#### Known Alfresco issues that impact Dynamic Extensions
```

> <details><summary>Alfresco 6.1 - wrong version of 'Commons annotations' used</summary>When using DE on Alfresco 6.1, it is possible that it fails to startup due to following error:

Reviewer: You can use a header in the `<summary>` I think

Author: I tried locally, didn't work...
(Which could also be an issue with my local markdown viewer / formatter)

* Remove the `jsr250-api` and `geronimo-annotation_1.0_spec` jars from the `WEB-INF/lib` folder of the Alfresco webapp.
* Install [this hotfix AMP](https://github.com/xenit-eu/alfresco-hotfix-MNT-20557) in your Alfresco distribution, 
which will overwrite the `jsr250-api` and `geronimo-annotation_1.0_spec` jars with empty jars.
</details>

## Documentation

> Please checkout the [documentation](documentation) for further instructions.

Reviewer 2: I think `[documentation](documentation#readme)` should work to automatically jump to the table of contents.
